### PR TITLE
WeightInfo for System, Timestamp, and Utility

### PR DESCRIPTION
**Pull request by shawntabrizi**
This PR implements the `WeightInfo` trait for Frame System and the Timestamp and Utility Pallets.

polkadot companion: https://github.com/paritytech/polkadot/pull/1606

## System

|                         | Before             | After                         | Notes             |
|-------------------------|--------------------|-------------------------------|-------------------|
| remark                  | 700_000            | 1_305_000                     |                   |
| set_heap_pages          | 1_500_000 + 1W     | 2_023_000 + 1W                |                   |
| set_changes_trie_config | 10_000_000 + 2W    | 10_026_000 + 1R + 2W          | +1R from `append` api |
| set_storage             | (600_000 + 1W) * i | (656_000 + 1W) * i            |                   |
| kill_storage            | (400_000 + 1W) * i | 4_327_000 + (478000 + 1W) * i | Constant?         |
| kill_prefix             | (850_000 + 1W) * i | 8_349_000 + (838000 + 1W) * i | Constant?         |
| suicide                 | 10_000_000         | 29_247_000                    | Big change?       |

## Timestamp

|             | Before              | After               | Notes                                                         |
|-------------|---------------------|---------------------|---------------------------------------------------------------|
| set         | 8_000_000 + 2R + 1W | 9_133_000 + 2R + 1W | One read lost due to whitelisting the whole storage item |
| on_finalize | 5_000_000           | 5915000   |      |

## Utility

|               | Before                     | After                      | Notes |
|---------------|----------------------------|----------------------------|-------|
| batch         | 15_000_000 + 1_000_000 * c | 16_461_000 + 1_982_000 * c |       |
| as_derivative | 3_000_000                  | 4_086_000                  |       |

### CODE REVIEW

File: frame/utility/src/default_weights.rs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! THIS FILE WAS AUTO-GENERATED USING THE SUBSTRATE BENCHMARK CLI VERSION 2.0.0-rc5
+
+#![allow(unused_parens)]
+#![allow(unused_imports)]
```
Comment (kianenigma) [inline frame/utility/src/default_weights.rs:21]: I think it is fine to add this to all files as well for now to make it easier.

**Comment by shawntabrizi** [comment]
> Weight values are same benchmarks re-executed with the automation right?

Yes, and they seem a little slower... but I have no reason to think anything wrong about them.

File: frame/timestamp/src/lib.rs [hunk 2]
```diff
@@ -161,17 +157,8 @@ decl_module! {
...
 		/// The dispatch origin for this call must be `Inherent`.
-		///
-		/// # <weight>
-		/// - `O(T)` where `T` complexity of `on_timestamp_set`
-		/// - 1 storage read and 1 storage mutation (codec `O(1)`). (because of `DidUpdate::take` in `on_finalize`)
-		/// - 1 event handler `on_timestamp_set` `O(T)`.
-		/// - Benchmark: 7.678 (min squares analysis)
-		///   - NOTE: This benchmark was done for a runtime with insignificant `on_timestamp_set` handlers.
-		///     New benchmarking is needed when adding new handlers.
-		/// # </weight>
```
Comment (thiolliere) [inline frame/timestamp/src/lib.rs:172]: Why removing all this ? I would still prefer keeping it, this is useful to check if benchmarks are missing some complexity, I would just remove
```
		/// - Benchmark: 7.678 (min squares analysis)
		///   - NOTE: This benchmark was done for a runtime with insignificant `on_timestamp_set` handlers.
		///     New benchmarking is needed when adding new handlers.
```

File: frame/utility/src/lib.rs [hunk 3]
```diff
@@ -145,7 +141,8 @@ decl_module! {
 		#[weight = (
 			calls.iter()
 				.map(|call| call.get_dispatch_info().weight)
-				.fold(15_000_000, |a: Weight, n| a.saturating_add(n).saturating_add(1_000_000)),
+				.sum::<Weight>()
```
Comment (thiolliere) [inline frame/utility/src/lib.rs:144]: I'm not sure about this sum, isn't it a wrapping add ?
In some extirnsic user can give an upper on some size of on-chain stuff, user could then just make upper bound being very high, this sum would wrap around and final weight will be just the add below

Comment (shawntabrizi) [reply frame/utility/src/lib.rs:144]: This is a good point. I did not consider this

**Comment by thiolliere** [comment]
> `set_changes_trie_config | 10_000_000 + 2W | 10_026_000 + 1R + 2W | +1R? (TODO CHECK)`

it is probably the deposit_log

> It doesnt know that the item is cleaned up with on_finalize

for timestamp it is not important, so writing some todo in the whitelist issue is fine to me.

**Comment by shawntabrizi** [comment]
Its not clear where the increased constants from `kill_storage` and `kill_prefix` come from, but I have run the benchmark multiple times, and it is consistent.

Maybe something to follow up later where we can optimize the pipeline in some ways.

**Comment by shawntabrizi** [comment]
bot merge

**Pull request merged**